brainiak/reprsimil/brsa.py

```diff
@@ -0,0 +1,1118 @@
+                    n_C, n_T, n_V, rank, GP_space, GP_inten, dist2,
+                    inten_diff2, space_smooth_range, inten_smooth_range,
+                    inten_weight),
+                method='BFGS', jac=True, tol=10**(-2 - 2 / n_iter * (it + 1)),
+                options={'xtol': 10**(-3 - 3 / n_iter * (it + 1)),
+                         'disp': self.verbose, 'maxiter': 6})
+
+            current_logSNR2[0:n_V - 1] = res_fitV.x[0:n_V - 1]
+            current_logSNR2[n_V - 1] = -np.sum(current_logSNR2[0:n_V - 1])
+
+            fitVchange = res_fitV.x[:n_V - 1] - param0_fitV[:n_V - 1]
+            norm_fitVchange = np.linalg.norm(fitVchange)
+
+            param0_fitV = res_fitV.x.copy()
+            logger.debug('norm of parameter change after fitting V: '
+                         '{}'.format(norm_fitVchange))
+            logger.debug('E[log(SNR2)^2]: {}'.format(
+                np.mean(current_logSNR2**2)))
+
+            # fit U
+
+            param0_fitU[0:n_l] = current_vec_U_chlsk_l_AR1
+            param0_fitU[n_l:n_l + n_V] = current_a1
+
+            res_fitU = scipy.optimize.minimize(
+                self._loglike_y_AR1_diagV_fitU, param0_fitU,
+                args=(XTX, XTDX, XTFX, YTY_diag, YTDY_diag, YTFY_diag,
+                      XTY, XTDY, XTFY, current_logSNR2, l_idx, n_C, n_T, n_V,
+                      rank),
+                method='BFGS', jac=True, tol=10**(-2 - 2 / n_iter * (it + 1)),
+                options={'xtol': 10**(-2 - 3 / n_iter * (it + 1)),
+                         'disp': self.verbose, 'maxiter': 6})
+            current_vec_U_chlsk_l_AR1 = res_fitU.x[0:n_l]
+            current_a1 = res_fitU.x[n_l:n_l + n_V]
+
+            fitUchange = res_fitU.x - param0_fitU
+            norm_fitUchange = np.linalg.norm(fitUchange)
+            logger.debug('norm of parameter change after fitting U: '
+                         '{}'.format(norm_fitUchange))
+            param0_fitU = res_fitU.x.copy()
+
+            # Debugging purpose. But it exceeds complexity limit
+            # if GP_space:
+            #     current_GP = res_fitV.x[n_V - 1:n_V - 1 + n_smooth]
+            #     logger.debug('current GP[0]: {}'.format(current_GP[0]))
+            #     logger.debug('gradient for GP[0]: {}'.format(
+            #         res_fitV.jac[n_V - 1]))
+            #     if GP_inten:
+            #         logger.debug('current GP[1]: {}'.format(current_GP[1]))
+            #         logger.debug('gradient for GP[1]: {}'.format(
+            #             res_fitV.jac[n_V]))
+            if np.max(np.abs(fitVchange)) < tolx and \
+                    np.max(np.abs(fitUchange)) < tolx:
+                break
+
+        Uest_chlsk_l_AR1_UV = np.zeros([n_C, rank])
+        Uest_chlsk_l_AR1_UV[l_idx] = current_vec_U_chlsk_l_AR1
+
+        est_cov_AR1_UV = np.dot(Uest_chlsk_l_AR1_UV, Uest_chlsk_l_AR1_UV.T)
+
+        opt_rho1_AR1_UV = 2 / np.pi * np.arctan(current_a1)
```

Comment:
Worth checking whether tan or scaled logit is a better transformation to unconstrained space (w.r.t. speed, quality of estimate, and the gradients). I thought logit was more standard for this (but maybe it's just stan).